CMakeLists.txt

```diff
@@ -0,0 +1,137 @@
+# Builds the libraries for Boost.Nowide
+# Automatically detects if build as standalone or as "part of boost" (with boost namespace)
+#
+# Options:
+# NOWIDE_INSTALL
+# NOWIDE_BUILD_TESTS
+# NOWIDE_SYSTEM_INCLUDE
+#
+# Created target: Boost::nowide
+#
+# When not using CMake to link against the shared library on windows define -DBOOST_NOWIDE_DYN_LINK
+
+cmake_minimum_required(VERSION 3.9)
+project(BoostNowide VERSION 1.0.0 LANGUAGES CXX)
+
+if(POLICY CMP0074)
+  cmake_policy(SET CMP0074 NEW)
+endif()
+
+if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+  set(is_root_project ON)
+  set(is_sub_project OFF)
+else()
+  set(is_root_project OFF)
+  set(is_sub_project ON)
+endif()
+
+# Find out if we are beeing build as standalone or boost version
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/include/nowide)
+  set(NOWIDE_STANDALONE ON)
+else()
+  set(NOWIDE_STANDALONE OFF)
+endif()
+
+option(NOWIDE_INSTALL "Install library" ${is_root_project})
+option(NOWIDE_BUILD_TESTS "Build unit tests" ${is_root_project})
+option(NOWIDE_SYSTEM_INCLUDE "Treat Boost.Nowide includes as system includes" ${is_sub_project})
+option(NOWIDE_WERROR "Treat warnings as errors" ${is_root_project})
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  set(warningFlags -Wall -Wextra -pedantic)
+  if(NOWIDE_WERROR)
+    list(APPEND warningFlags -Werror)
+  endif()
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
+  set(warningFlags -Wall)
+  if(NOWIDE_WERROR)
+    list(APPEND warningFlags -Werror)
+  endif()
+elseif(MSVC)
+  set(warningFlags /W4)
+  if(NOWIDE_WERROR)
+    list(APPEND warningFlags /WX)
+  endif()
+endif()
+
+# Make sure all binarys (especially exe/dll) are in one directory for tests to work
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+
+add_library(boost_nowide)
+set_target_properties(boost_nowide PROPERTIES
+  DEFINE_SYMBOL BOOST_NOWIDE_DYN_LINK
+  POSITION_INDEPENDENT_CODE ON
+  EXPORT_NAME nowide
+)
+add_library(Boost::nowide ALIAS boost_nowide)
+target_compile_options(boost_nowide PRIVATE ${warningFlags})
+target_compile_definitions(boost_nowide PUBLIC BOOST_NOWIDE_NO_LIB)
+
+if(NOWIDE_STANDALONE)
+  target_compile_features(boost_nowide PUBLIC cxx_std_11)
+else()
+  # Default boost libs are static on windows and dynamic on linux
+  if(WIN32 AND "${Boost_USE_STATIC_LIBS}" STREQUAL "")
+    set(Boost_USE_STATIC_LIBS ON)
+  endif()
+  find_package(Boost 1.55 REQUIRED)
```

Comment:
When you're built as part of the Boost superproject (https://github.com/boostorg/boost/blob/develop/CMakeLists.txt) you shouldn't issue `find_package(Boost)`.

My current "model" library is https://github.com/boostorg/timer/blob/develop/CMakeLists.txt, https://github.com/boostorg/timer/blob/develop/.travis.yml#L149-L154.

Reply:
Oh, great!

Ok so I guess I can simply put that inside an `if` for `BOOST_ENABLE_CMAKE`? I definitely need it to work with a (b2) installed boost for development and consumption as git submodule, that's why the find_package.

If you see anything else, feel free to comment.

A tip in return: You can simplify https://github.com/boostorg/timer/blob/0ed7b642a9a6ddddc7d838a2edf6d9a70692a16e/CMakeLists.txt#L25-L30 to something like https://github.com/boostorg/nowide/pull/27/files/b7141f8757d515f187cb9fd65a46bb2b4a94aca8#diff-af3b638bc2a3e6c650974192a53c7291R61-R65. CMake already has some logic for shared libs definitions. There won't be a static definition but e.g. `BOOST_TIMER_STATIC_LINK` is never used (or is it?). IMO the `*_DYNAMIC_LINK` is enough anyway.

Reply:
The correct check would be for either BOOST_SUPERPROJECT_VERSION or BOOST_SUPERPROJECT_SOURCE_DIR being defined. BOOST_ENABLE_CMAKE is temporary; at some point when we officially support building with CMake we'll remove it.

Defining options when inside the superproject will probably be declared bad practice at some point (because if 140 libraries define 4 options each, the user will have a lot to scroll through), but this may be a lost battle.

My current thinking is that we should install the config files to `cmake/{PROJECT_NAME}-{PROJECT_VERSION}`, same as the current `b2 install`. But I'm not sure what libraries that use their own versioning ought to do. I certainly intend to require exact version match in all my `find_dependency` calls, but if nobody depends on Nowide, this won't matter for you.

We'll figure these things out with time, I guess. For now, it's important for libraries to be good citizens and not break the superproject, we'll see about the rest.

I didn't know about DEFINE_SYMBOL, thanks.

Reply:
> My current thinking is that we should install the config files to cmake/{PROJECT_NAME}-{PROJECT_VERSION}, same as the current b2 install.

Does it? How would a find_package be able to find that? When installing to `cmake/PROJECT_NAME` it will be found by `find_package(PROJECT_NAME)`. Or is there something that extends the search-prefixes?

> But I'm not sure what libraries that use their own versioning ought to do

It seems for CMake this does not matter unless we to the same name mangling that b2 does: You won't be able to install a library with 2 different versions into the same prefix. And per prefix the search is perfectly fine. When you provide a "VersionConfig", you can even have all those prefixes in your PREFIX_PATH and the search will pick the first matching version while discarding the rest. So I think the way to go there would be to install into e.g. `/opt/boost_<VERSION` and users add each of them to their CMAKE_PREFIX_PATH.

But well, this gonna take some time.

> Defining options when inside the superproject will probably be declared bad practice at some point

I tend to disagree. When those options have reasonable defaults, then why not? They should be "standardized" though. That's how I chose mine. Because if you have to edit a CMakeLists or so to not build the tests or not install... that would be very inconvenient. And having an option to disable Werror is also a good idea (I'm pro leaving that on and building with highest warning settings, catches errors early and leaves less log clutter for users)

Reply:
> How would a find_package be able to find that?

It does. Try it.

> When those options have reasonable defaults, then why not?

Well the obvious reason is that 4 * 140 = 560. Even if every option by itself makes perfect sense, should a user really be expected to wade through 560 options? Doesn't seem very practical, especially when most of these will overlap so you'll want to set, say, 80 of them at once because they do the same thing.

Reply:
Ok nice, how did you know? It is not mentioned in the docu... Well, if using that then we'd need to suffix the version to the libraries or the configs will pick the last installed version

That's why I said a standardized set and useful defaults. I imagine 2 use cases:
- installing all of boost, either with all tests or without all tests
- testing a single library -> disable all tests + installation, enable for the single library

Hence the default value for the boost super build powered run should be the same standardized variables with simple Boost_ prefix: `option(Boost_NOWIDE_INSTALL "Install library" ${Boost_INSTALL})`

Btw: In https://github.com/boostorg/boost/blob/develop/CMakeLists.txt we should use a "good" project name to avoid setting variables with different prefixes. If "Boost" is already taken by the b2 build, then BoostSuper? BoostRoot? BOOST_SUPERPROJECT? (But that is to screamy). Or just keep it with "Boost" and one can check for "Boost_SOURCE_DIR" to see if build as part of it.

Reply:
Have you seen https://gitlab.otris.de/vendor/boost_cmake? Might be good for inspiration but it doesn't seem to be that good...